reference/preconditioner/isai_kernels.cpp

```diff
@@ -0,0 +1,223 @@
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/preconditioner/isai_kernels.hpp"
+
+
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/matrix/csr_builder.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The Isai preconditioner namespace.
+ *
+ * @ingroup isai
+ */
+namespace isai {
+
+
+template <typename ValueType, typename IndexType, typename Callable>
+void generic_generate(std::shared_ptr<const DefaultExecutor> exec,
+                      const matrix::Csr<ValueType, IndexType> *mtx,
+                      matrix::Csr<ValueType, IndexType> *inverse_mtx,
+                      Callable trs_solve)
+{
+    /*
+    Consider: aiM := inverse_mtx; M := mtx
+    I := Identity matrix
+    e(i) := unit vector i (containing all zeros except for row i, which is one)
+    S := Sparsity pattern of the desired aiM
+    S(i) := Sparsity pattern of row i of aiM (Set of non-zero columns)
+    D(i) := L[S(i), S(i)]
+    aiM := approximate inverse of M
+
+    Target: Solving (aiM * M = I)_{S} (aiM * M = I for the sparsity pattern S)
+    aiM[i, :] * D(i) = e(i)^T
+    <=> D(i)^T * vL[i, :]^T = e(i)   =^ Triangular system (Trs)
+    Solve Trs, fill in aiM row by row (coalesced access)
+    */
+    const auto size = mtx->get_size();
+    const auto m_row_ptrs = mtx->get_const_row_ptrs();
+    const auto m_cols = mtx->get_const_col_idxs();
+    const auto m_vals = mtx->get_const_values();
```

Comment:
To maximize confusion: 
```suggestion
    const auto m_col_ptrs = mtx_csc->get_const_row_ptrs();
    const auto m_rows = mtx_csc->get_const_col_idxs();
    const auto m_vals = mtx_csc->get_const_values();
```
This is of course a matter of opinion, this suggestion aims at reducing the amount of abstractions (or hidden transpositions) you need to keep in mind to understand the code.